server/src/main/scala/org/http4s/server/ServerBuildable.scala

```diff
@@ -0,0 +1,141 @@
+  */
+trait ServerBuildable[F[_], A] {
+
+  /** Returns the backend as a resource.  Resource acquire waits
+    * until the backend is ready to process requests.
+    */
+  def resource(a: A): Resource[F, Server]
+
+  // Abstract Settings
+
+  /** Set the http host for the server. */
+  def withHttpHost(a: A)(host: String): A
+
+  /** Set the http port for the server. */
+  def withHttpPort(a: A)(port: Int): A
+
+  /** Set the [[HttpApp]] to serve on this server. */
+  def withHttpApp(a: A)(app: HttpApp[F]): A
+
+  // final
+
+  /** Returns the backend as a single-element stream.  The stream
+    * does not emit until the backend is ready to process requests.
+    * The backend is shut down when the stream is finalized.
+    */
+  final def stream(a: A): Stream[F, Server] = Stream.resource(resource(a))
+
+  /** Returns an effect that allocates a backend and an `F[Unit]` to
+    * release it.  The returned `F` waits until the backend is ready
+    * to process requests.  The second element of the tuple shuts
+    * down the backend when run.
+    *
+    * Unlike [[resource]] and [[stream]], there is no automatic
+    * release of the backend.  This function is intended for REPL
+    * sessions, tests, and other situations where composing a
+    * [[cats.effect.Resource]] or [[fs2.Stream]] is not tenable.
+    * [[resource]] or [[stream]] is recommended wherever possible.
+    */
+  final def allocated(a: A)(implicit F: Bracket[F, Throwable]): F[(Server, F[Unit])] = resource(
+    a).allocated
+
+  /** Bind the server to a given port and host. */
+  final def bindHttp(a: A)(port: Int = defaults.HttpPort, host: String = defaults.Host): A =
+    withHttpPort(withHttpHost(a)(host))(port)
+
+  /** Bind the server to localhost */
+  final def bindLocal(a: A)(port: Int): A = bindHttp(a)(port, defaults.Host)
+
+  /** Bind the server to localhost, on some dynamically determined free port. */
+  final def bindAny(a: A)(host: String = defaults.Host): A = bindHttp(a)(0, host)
+
+  /** Runs the server as a Stream that emits only when the terminated signal becomes true.
+    * Useful for servers with associated lifetime behaviors.
+    */
+  final def serveWhile(a: A)(
+      terminateWhenTrue: Signal[F, Boolean],
+      exitWith: Ref[F, ExitCode]): Stream[F, ExitCode] =
+    Stream.resource(resource(a)) *> (terminateWhenTrue.discrete
+      .takeWhile(_ === false)
+      .drain ++ Stream.eval(exitWith.get))
+
+  /** Runs the server as a process that never emits.  Useful for a server
+    * that runs for the rest of the JVM's life.
+    */
+  final def serve(a: A)(implicit F: Concurrent[F]): Stream[F, ExitCode] =
+    for {
+      signal <- Stream.eval(SignallingRef[F, Boolean](false))
+      exitCode <- Stream.eval(Ref[F].of(ExitCode.Success))
+      serve <- serveWhile(a)(signal, exitCode)
+    } yield serve
+}
+
+object ServerBuildable {
+
+  // TODO: Generate syntax with simulacrum? It's a two parameter class so I'm
+  // not certain that is viable.
```

Comment:
I think that would also cause Scala 3 problems.

Reply:
If we went the typeclass route, I was thinking of using https://github.com/typelevel/simulacrum-scalafix, which I think would be fine for Scala 3, though I don't know about the higher arity class support.